docs/source/developer_guide/generator-runner.rst

```diff
@@ -0,0 +1,358 @@
+.. _generator_runner:
+
+
+How to use ``datalad``'s runner with Generator Protocols
+********************************************************
+
+``datalad_next`` executes a lot of subprocesses to get its work done.
+Subprocess execution is done by the runner class, i.e. ``datalad_next.runner.ThreadedRunner`` and a small shell around it: ``datalad_next.runner.Runner`` (we refer to the latter as "runner" throughout the rest of the document).
+
+This document is intended for ``datalad``-developers who want to understand how the runner code works and how they can use it to efficiently execute subprocesses or to build higher level tools, e.g. batch-command classes.
+
+
+General functionality
+=====================
+The runner will execute a subprocess and communicate with it. Communication means that it might send input the the stdin-descriptor of the subprocess, if that is specified, and read output from the stderr- and stdout-descriptors of the subprocess, if that is specified.
+
+All input to the subprocess is directly written the stdin-descriptor of the subprocess-object (which is usually the write end of a pipe-like structure).
+The output of the subprocess is passed to an instance of a `protocol`-class in the thread that executed the ``run``-method of the runner. The protocol classes are modelled after the class ``asyncio.protocol.SubprocessProtocol``. They provide callbacks for the following events:
+
+
+- data received from stdout or stderr of the subprocess: ``pipe_data_received``
+- data connection lost: ``pipe_connection_lost``
+- connection to subprocess established: ``connection_made``
+- connection to subprocess closed: ``connection_lost``
+- subprocess exited: ``process_exited``
+- timeout occurred: ``timeout``
+
+When the subprocess is executed the runner will invoke the appropriate callbacks.
+
+``datalad_next`` provides some pre-defined protocol-classes for common use cases, e.g. ``datalad_next.runners.StdOutErrCapture``, which executes a subprocess and returns a dictionary with the ``stdout``- and ``stderr``-output, and the `return code` (also referred to as `exit status` ) of the subprocess.
```

Comment:
This sounds like a good place to mention all pre-defined protocol classes to give a quick overview?

(also found a stray space)
```suggestion
``datalad_next`` provides some pre-defined protocol-classes for common use cases, e.g. ``datalad_next.runners.StdOutErrCapture``, which executes a subprocess and returns a dictionary with the ``stdout``- and ``stderr``-output, and the `return code` (also referred to as `exit status`) of the subprocess.
```

Reply:
> This sounds like a good place to mention all pre-defined protocol classes to give a quick overview?

That is a good idea. I will add that. Depending on how bulky it gets, I might just mention the "top" ("top" as in most-used)  protocols.

Reply:
On  a second thought. At the position mentioned above the generator-mode is not yet described. I will therefore add the list at the end the section "General functionality", just before the section "Programming examples".

Reply:
done in 43fc79a4e5cf3067fe04d1bd8e9c5787e8bb2a1d